src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -873,6 +923,92 @@ NodeIsFieldStore(Node *node)
 }
 
 
+/*
+ * MergeQuerySupported does check for a MERGE command in the query, if it finds
+ * one, it will verify the below criteria
+ * - Supported tables and combinations in ErrorIfMergeHasUnsupportedTables
+ * - Distributed tables requirements in ErrorIfDistTablesNotColocated
+ * - Checks target-lists and functions-in-quals in TargetlistAndFunctionsSupported
+ */
+static DeferredErrorMessage *
+MergeQuerySupported(Query *originalQuery,
+					PlannerRestrictionContext *plannerRestrictionContext)
+{
+	/* For non-MERGE commands it's a no-op */
+	if (!QueryHasMergeCommand(originalQuery))
+	{
+		return NULL;
+	}
+
+	List *rangeTableList = ExtractRangeTableEntryList(originalQuery);
+	RangeTblEntry *resultRte = ExtractResultRelationRTE(originalQuery);
+
+	/*
+	 * Fast path queries cannot have merge command, and we prevent the remaining here.
+	 * In Citus we have limited support for MERGE, it's allowed only if all
+	 * the tables(target, source or any CTE) tables are are local i.e. a
+	 * combination of Citus local and Non-Citus tables (regular Postgres tables)
+	 * or distributed tables with some restrictions, please see header of routine
+	 * ErrorIfDistTablesNotColocated for details.
+	 */
+	DeferredErrorMessage *deferredError =
+		ErrorIfMergeHasUnsupportedTables(originalQuery,
+										 rangeTableList,
+										 plannerRestrictionContext);
+	if (deferredError)
+	{
+		return deferredError;
+	}
+
+	Oid resultRelationId = resultRte->relid;
+	deferredError =
+		TargetlistAndFunctionsSupported(resultRelationId,
+										originalQuery->jointree,
+										originalQuery->jointree->quals,
+										originalQuery->targetList,
+										originalQuery->commandType,
+										originalQuery->returningList);
+	if (deferredError)
+	{
+		return deferredError;
+	}
+
+	#if PG_VERSION_NUM >= PG_VERSION_15
```

Comment:
>>and also makes sure we do not get into any of the merge related code in PG 14 ?

The very first line of code does that

        /* function is void for pre-15 versions of Postgres */
        #if PG_VERSION_NUM < PG_VERSION_15
        return false;
        #else
Let me see, how to rearrange so that it compiles on PG 14 too

Reply:
Moved the macro to the very top of the function, if we make the function header too come under the purview of the macro, compile will fail on PG 14 and below versions, and we have to add another macro at the caller. This should keep it simple.